File: src/main/java/com/warren/bt/SuccessRate.java
package com.warren.bt;

import java.util.Random;
import java.util.function.Supplier;

public class SuccessRate extends Decorator {
  private final Supplier<Float> successRateSupplier;
  private final Random random;

  public SuccessRate(Node child, float successRate) {
    this(child, () -> successRate);
  }

  public SuccessRate(Node child, Supplier<Float> successRateSupplier) {
    super(child);
    this.successRateSupplier = successRateSupplier;
    this.random = new Random();
  }

  @Override
  public Status tick() {
    if (child == null) {
      return Status.FAILURE;
    }

    float currentSuccessRate = Math.max(0f, Math.min(100f, successRateSupplier.get()));

    float normalizedRate = currentSuccessRate / 100f;

    if (random.nextFloat() > normalizedRate) {
      return Status.FAILURE;
    }

    return child.tick();
  }
}

File: src/main/kotlin/com/warren/bt/SuccessRate.kt
package com.warren.bt

import kotlin.random.Random

/**
 * Probabilistic gate: each tick, with probability [chance], tick the child; otherwise FAIL
 * immediately.
 */
class SuccessRate(private val chance: Float) : Decorator("SuccessRate") {
  override fun tick(): Status {
    if (chance <= 0f) return Status.FAILURE
    if (chance >= 1f) return child?.tick() ?: Status.FAILURE
    if (Random.nextFloat() > chance) return Status.FAILURE
    return child?.tick() ?: Status.FAILURE
  }
}
